libtenzir/builtins/operators/deduplicate.cpp

```diff
@@ -122,33 +136,143 @@ class deduplicate_operator final : public crtp_operator<deduplicate_operator> {
+            "failed to deduplicate due to unmatched type extractor")
+            .note("`{}` matched no fields in schema `{}`", field, schema.name())
+            .emit(diag);
+          return std::nullopt;
+        }
+        missing_fields.emplace_back(field);
       }
     }
-    std::sort(indices.begin(), indices.end());
-    indices.erase(std::unique(indices.begin(), indices.end()), indices.end());
-    return indices;
+    if (missing_fields.empty()) {
+      // Every field in `cfg_.fields` was found in `slice.schema()`.
+      // Clean up the indices and return.
+      std::ranges::sort(indices);
+      {
+        auto uniq = std::ranges::unique(indices);
+        indices.erase(uniq.begin(), uniq.end());
+      }
+      return cached_projection{
+        .indices = std::move(indices),
+        .transformation_factory = {},
+      };
+    }
+    // Some fields were missing.
+    // Construct a `transformation_factory`, that'll add these fields in, with a
+    // value of `null`.
+    TENZIR_ASSERT_EXPENSIVE(std::unordered_set<std::string>(
+                              missing_fields.begin(), missing_fields.end())
+                              .size()
+                            == missing_fields.size());
+    const auto& layout = caf::get<record_type>(schema);
+    auto transformation_callback =
+      [missing_fields](int64_t rows, struct record_type::field input_field,
+                       std::shared_ptr<arrow::Array> input_array) {
+        auto result = std::vector<
+          std::pair<struct record_type::field, std::shared_ptr<arrow::Array>>>{};
+        result.reserve(missing_fields.size() + 1);
+        result.emplace_back(std::move(input_field), std::move(input_array));
+        for (const auto& missing_field : missing_fields) {
+          auto builder
+            = null_type::make_arrow_builder(arrow::default_memory_pool());
+          {
+            const auto append_result = builder->AppendNulls(rows);
+            TENZIR_ASSERT(append_result.ok(), append_result.ToString().c_str());
+          }
+          using field_type = struct record_type::field;
+          result.emplace_back(field_type{missing_field, type{null_type{}}},
+                              builder->Finish().ValueOrDie());
+        }
+        return result;
+      };
+    auto transformation_factory
+      = [num_fields = layout.num_fields(),
+         transformation_callback = std::move(transformation_callback)](
+          const table_slice& slice) -> indexed_transformation {
+      return {
+        {num_fields - 1},
+        [rows = slice.rows(),
+         &transformation_callback](struct record_type::field input_field,
+                                   std::shared_ptr<arrow::Array> input_array) {
+          return transformation_callback(static_cast<int64_t>(rows),
+                                         std::move(input_field),
+                                         std::move(input_array));
+        },
+      };
+    };
```

Comment:
I have a feeling this doesn't handle nested fields correctly. Can you confirm/debunk this?

Reply:
Oh right it doesn't, as this is semi-directly pulled from `put`/`extend`. I added a `flatten` call before any projection logic, which should make this work.

Reply:
Nested fields are now handled correctly.